docs/design/sql/16-partition-aware-client.md

```diff
@@ -0,0 +1,48 @@
+# Partition-aware client
+
+If we want SQL to be the primary interface to Hazelcast, the basic map get/put
+operations using SQL have to be equally fast as their native counterparts. We’ve
+implemented shortcut evaluation for the queries where the WHERE clause contains
+`__key=?` - we don’t start a job in this case, but directly execute
+`IMap.get()`. It’s also implemented for INSERT/SINK, UPDATE and DELETE
+statements.
+
+However, the operation is still much slower than the `IMap` counterparts. The
+main reason is that the client sending a query picks a random member to execute
+the query, unaware of the partition owner.
+
+The original idea to fix this was to use a kind of client-side plan cache. Along
+with the query result, the cluster will send a plan to the client telling it
+that the next time it sees that query, instead of sending it as a query it
+should execute IMap.get directly. There are two issues with this approach:
+
+1. cache invalidation: the member will have to keep track what plans were cached
+   by which client, and will need to invalidate those 
+
+2. client-side complexity: each type of operation will have to be implemented by
+   each client. It’s not just simple `get`, but `get` with projections, additional
+   filters, and also `put` for INSERT, and entry processor for UPDATE, which might
+   not be possible to implement on non-java client.
+
+Another option to implement this is much simpler on the client. Along with a
+result, the cluster will send from which argument to derive the partition key,
+and submit the query to the owner of that key. If the argument implements
+`PartitionAware`, that should be taken into account. This approach doesn't have
+any of the above disadvantages. Plan cache will remain only on members, and the
+client will remain simple.
```

Comment:
Can't we use prepared statements for this? It would work like:
1. The user calls `client.getSql().createPreparedStatement(sql)`
2. SQL service sends a message to the member (let's call it returnPartitionAwareKey) via the client, before returning the prepared statement
3. The member finds out the partition aware field, and sends its index (in the SQL), or `-1` if there's no partition aware field back. (We may also return a "plan ID" if something like that makes sense)
4. SQL service sets the partitionAwareIndex in the prepared statement and returns the prepared statement to the user.
5. Whenever the user runs `execute` on the prepared statement, the argument at the partitionAwareIndex is used to calculate the partitionID.

Many database drivers have prepared statements, so I would expect most SQL users are familiar with them.

This approach solves a few issues:
1. The first execution is as fast as others.
2. The behaviour is very predictable, the user knows this is an optimized way of running a query.
3. No "global" cache on the client-side is required, which makes the client side less complicated and also doesn't require factoring in the eviction size, etc.

Reply:
However, does not this require two requests for each SQL statement? So, are users expected to reuse the prepared statement? If so, we should make sure users don't create a new prepared statement for each query whose SQL strings are the same. Maybe we can log a warning by detecting that. However, it will be impossible to detect that without a cache. 

Other than this, I think this approach makes sense.

Reply:
> However, does not this require two requests for each SQL statement? So, are users expected to reuse the prepared statement?

The user is expected to create the prepared SQL once and re-use it as many times as necessary afterwards.

> If so, we should make sure users don't create a new prepared statement for each query whose SQL strings are the same. > Maybe we can log a warning by detecting that.

IMO that's not necessary. It's common knowledge that the prepared statement should be created once and reused.

Reply:
This proposal makes the first execution slower for queries that do not a have partition argument. It's beneficial if the resultset is large, because it's not the whole data that's shipped twice, but if the result-set is small, the benefit is small or negative (typically, the client-to-member RTT is larger than member-to-member RTT).

Second, I think it's common that PreparedStatements are  not reused in user code. Very often, they are created just to execute a single query, because it's the way to execute any query with parameters. For example, PG even caches the prepared statements transparently (the `borrowQuery` call [here](https://github.com/pgjdbc/pgjdbc/blob/0ec7e15fe41170d657237b36add83eaf334d70a3/pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java#L88)).

So I don't think this is a clear win. Even worse, it slows down many common use cases. If we're worried about the cache size, we can use a lower default and allow the user to configure it. I'll elaborate that in the TDD.

Reply:
I've researched prepared queries again, and as far as I see reusing prepared statements is the best practice.

If we expect creating one-off prepared statements will be common, and it's OK to not hit the correct partition with the first query (like the proposal), then the `execute` method of the prepared statement would just:
1. First time it is invoked, it sends the query to a random member, since partition aware index is not known,
2. The member replies back with the partition aware index (like the proposal) and prepared query saves it locally,
3. The next time, it uses the partition aware index.

This solves the problem with one-off queries calling the member-side twice while not requiring a global cache

Reply:
> This proposal makes the first execution slower for queries that do not a have partition argument. It's beneficial if the resultset is large, because it's not the whole data that's shipped twice, but if the result-set is small, the benefit is small or negative (typically, the client-to-member RTT is larger than member-to-member RTT).

Maybe if a user uses a prepared statement for the SQL that does not have a partition argument, we can log a warning. Moreover, we can document prepared statements should only be used when a parameter is partition aware. 

--- 

I think another approach suggested by @ihsandemir was to let users specify the partition aware argument index. @viliam-durina WDYT about it?

Reply:
I remember talking to advanced customers who likes to know which server their requests are being processed and they write logic against such information. Hence, I think that giving the freedom to the user for choosing the partition parameter makes sense for the users. This kind of API also will let a lot other use cases possible which gives total control to the users.

Reply:
Caching the partition argument index in the prepared statement is definitely an option to avoid the global cache. It's essentially reducing the scope of the cache, so it's a trade-off between cache misses and cache size and maintenance cost. If the user _always_ reuses the PStmt, then the global cache has no benefit. But I'm still not sure about this. I don't see the global cache cost as a large burden, as explained in the TDD.

Another question is whether other languages have a notion of a prepared statement like Java does. I worked only with PHP (very long ago), and it didn't. ODBC has it. And our native API doesn't have it.

>Maybe if a user uses a prepared statement for the SQL that does not have a partition argument, we can log a warning.

It is definitely not an error, or a problem worth warning about, if a query doesn't have partition argument index. In fact, most queries don't have it and can't possibly have it.

Regarding the `useServerHostingData`, it's doable, but I don't like it. The method name implies that the engine won't check members other than the specified one, which technically is a filtering condition. Such a condition should be in the query. So even if your query is `select * from my_map`, if you use this option, you'll not see all the entries from the map.

SQL is declarative. You specify _what_ you want, and the engine figures out _how_ to get it for you. For example, you cannot request the engine to use a particular index; the engine figures that out itself. You just create the index, and then you can check using `EXPLAIN` that it got used. If you drop or alter the index, your query will continue to work, though it might take 1000x times longer to complete without the index. I personally must say that I'm not a fan of this approach. It works for simple cases, but in more complex cases the engine screws it. As the optimizers get smarter, more and more often they screw things up. I worked mostly with PG and Oracle, and I lacked the optimizer hints in PG dearly. My favorite rule in oracle was `rule`, which meant to use the much simpler rule-based optimizer, but that optimizer was much more stable and reliable. Oracle even has some "plan management" to tackle the issue of query performance changes in face of changing statistics and stuff, but I didn't use that.

Reply:
`So even if your query is select * from my_map, if you use this option, you'll not see all the entries from the map.`
I do not agree with this, the query will still go to the member the the user chose and at the server side you will parse and retrieve the datas from other servers as well and return to the user.

Please do NOT misunderstand the method name. It does NOT mean that the engine will not check other members it needed. Think of this as a HINT for the query. We can also name it as `query.serverHint(Object);`

Reply:
A hint makes more sense. The SQL syntax for hint is a comment right after the SELECT keyword, whose first character is `+`, e.g.:
```sql
SELECT /*+ part_arg(1) */ *
FROM my_map
WHERE __key=?
```
But such a hint isn't processed on the client (client doesn't parse the SQL string), so it's useless for us now.

Also remember that a custom method is unlikely to be used. For example, you can't use it if you're using JDBC or another standardized API. But even a SQL hint is typically unusable if you're using some ORM framework.

Reply:
@viliam-durina I think removing the cache simplifies the design quite a bit on the client-side. Having a section dedicated to the cache design in the TDD speaks for itself :)

I would say prepared statements is common, Go also has that in its standard SQL spec. Python doesn't have it in its standard SQL spec, but some drivers (like the MySQL) driver implement that.

Reply:
Btw, we don't need a separate prepared statement concept for this this idea to work with our API (since we don't need `.createPreparedStatement` -like thing anymore), the user would just do as usual:
```java
SqlStatement stmt = new SqlStatement("SELECT ...");
stmt.setParameters(Collections.singletonList("p1"));
SqlResult res = client.getSql().execute(stmt);
// the second run may be partition aware
stmt.setParameters(Collections.singletonList("p2"));
res = client.getSql().execute(stmt);
```

Reply:
@yuce Yeah, we could use `SqlStatement` as a prepared statement, but the contract will have to change:
```java
/**
 * Definition of an SQL statement.
 * <p>
 * This object is mutable. Properties are read once before the execution is started.
 * Changes to properties do not affect the behavior of already running statements.
 */
public final class SqlStatement 
...
```

It explicitly says that you can reuse it after submitting a query. We still could use it as a cache for the partition arg index, it will work if somebody doesn't use it for multiple statements. If he does, the optimization will be broken, but no error will be thrown and results will be correct. We also have the `execute(String sql, Object ... args)` method - if using it, the optimization will not work, because it creates a new SqlStatement each time, so we probably should deprecate this method.

But I'm still not sure about the whole idea of a local cache like this. Of course, it depends on the usage pattern, whether you reuse the prepared stmts or not, but I see drivers having caches even for PreparedStatement objects themselves - if you create a new prepared stmt for the same sql text, they'll reuse some state. I don't see the need for global cache as a major complication, a custom cache was pretty easy to implement in Java.

Reply:
> It explicitly says that you can reuse it after submitting a query. We still could use it as a cache for the partition arg index, it will work if somebody doesn't use it for multiple statements.

If the user changes the SQL on a statement, we can just reset the partitionID so it would work just like a new statement, so no optimization. But I think this is quite an acceptable behaviour on the user side.

> I see drivers having caches even for PreparedStatement objects themselves - if you create a new prepared stmt for the same sql text, they'll reuse some state

Some drivers (like MySQL) seem to send the prepared statement to the server-side upon creation, so I can see why they would cache the prepared statements. But in our API, `SqlStatement` is very cheap to create, so probably it doesn't worth to cache them.

Reply:
>Some drivers (like MySQL) seem to send the prepared statement to the server-side upon creation, so I can see why they would cache the prepared statements. But in our API, SqlStatement is very cheap to create, so probably it doesn't worth to cache them.

That's true, but the bigger issue is the `SqlService.execute` variants that don't use `SqlStatement`. These show that `SqlStatement` was designed as a simple DTO. Do we have analogous issue in other HZ clients?

Reply:
> Also remember that a custom method is unlikely to be used. For example, you can't use it if you're using JDBC or another standardized API. 

Does JDBC drive even use smart client or unisocket client? partition aware client is useless for unisocket, right?